### PR TITLE
Migrate kev2 to feature charts

### PR DESCRIPTION
**Pull request by StrongMonkey**
This PR migrates kev2 install from system-chart to feature charts with ability to support on-demand install and min version. 

On-demand install: operator will only get installed if kev2 cluster is created.
Min version:  https://github.com/rancher/rancher/issues/30934

~Note: This PR relies on a fork of helm to add ability to adopt resources from helm v2 in order to handle upgrade scenario https://github.com/StrongMonkey/helm/commits/v3.5.4-rancher. I will try to get this into shell images.~ Helm changes are merged and included in rancher/shell:v1.0.7

https://github.com/rancher/rancher/issues/31592

### CODE REVIEW

**Comment by StrongMonkey** [comment]
Assigning forks that worked on kev2 for reviews. In case you found something that needs to be addressed.

**Review by thedadams — COMMENTED** [review]
Looks good. Will have conflicts with https://github.com/rancher/rancher/pull/32900

Also, can remove the following:
https://github.com/rancher/rancher/blob/master/pkg/controllers/management/clusteroperator/utils.go#L53
https://github.com/rancher/rancher/blob/master/pkg/controllers/management/clusteroperator/utils.go#L59
https://github.com/rancher/rancher/blob/master/pkg/controllers/management/clusteroperator/utils.go#L189

**Comment by cmurphy** [comment]
And also these:

https://github.com/rancher/rancher/blob/113c35317ea3524a8d6e9da6659c772377d33f2b/pkg/controllers/management/aks/aks_cluster_handler.go#L384-L388
https://github.com/rancher/rancher/blob/113c35317ea3524a8d6e9da6659c772377d33f2b/pkg/controllers/management/eks/eks_cluster_handler.go#L484-L488
https://github.com/rancher/rancher/blob/113c35317ea3524a8d6e9da6659c772377d33f2b/pkg/controllers/management/gke/gke_cluster_handler.go#L414-L418

**Comment by StrongMonkey** [comment]
@thedadams I will rebase on your changes if your PR gets merged earlier.

**Comment by StrongMonkey** [comment]
@thedadams @cmurphy Unused functions are removed.

**Review by thedadams — APPROVED** [review]

**Review by rmweir — COMMENTED** [review]
Looking through the install code it looks like this will try to keep the charts up to date with the latest version. What if the rancher version in not compatible with the latest charts version? Will it try to make sure rancher has the latest compatible version even if perhaps it cannot use the latest? If it does do this then LGTM after conflicts have been resolved.

**Comment by StrongMonkey** [comment]
@rmweir There is a minimal version setting for each operator so that we make sure for upgrading setup, if the current operator version meets the min version of the setting, it will not try to upgrade to the latest version. The behavior is defined in here #30934. This is to prevent that we don't intentionally upgrade user's operator if min version is satisfied.

But if there is a fresh install it will always try to get the latest version, this is not a one-operator-version-per-rancher-version model(2.6.0 -> 1.1.0, 2.6.1 -> 1.1.1), as it will always try to install the latest version of the chart in feature charts(so 2.6.x -> latest version of operator). But if there is a newer rancher-operator version that won't work with old rancher version, we can use rancher min/max version on catalog to filter out the latest charts. @cbron Thoughts?

**Comment by StrongMonkey** [comment]
Also there is also a point that for fresh install, user always install the latest version which should get the most compatiable version of operator at the time of the install.

**Review by rmweir — CHANGES_REQUESTED** [review]
@StrongMonkey do have parity with the way KEv2 currently works, it should always install the latest compatible version.

File: pkg/settings/setting.go
Comment (cbron) [inline pkg/settings/setting.go:118]: Also curious about this part, would we need to update every time ? Older version used `LatestAvailableTemplateVersion` to grab newest one. 

edit: 
Oh I missed your comment a few days back: 

> There is a minimal version setting for each operator so that we make sure for upgrading setup, if the current operator version meets the min version of the setting, it will not try to upgrade to the latest version. The behavior is defined in here #30934. This is to prevent that we don't intentionally upgrade user's operator if min version is satisfied.

> But if there is a fresh install it will always try to get the latest version, this is not a one-operator-version-per-rancher-version model(2.6.0 -> 1.1.0, 2.6.1 -> 1.1.1), as it will always try to install the latest version of the chart in feature charts(so 2.6.x -> latest version of operator). But if there is a newer rancher-operator version that won't work with old rancher version, we can use rancher min/max version on catalog to filter out the latest charts.

So to confirm we can use min/max to filter this ? Ideally we would always want the KEv2 operators to have the latest compatible chart for their version line. 
So 2.6.4 gets latest release in release-2.6 branch that is compatible with 2.6.4 (maybe newer one exists but min is 2.6.5). 
If we have to specify that in env vars that is possible, but not the best process. I'm surprised fleet uses this model, I know we don't want the _actual_ latest chart, but latest that uses min/max would be fine no ?

File: package/Dockerfile
Comment (cbron) [inline package/Dockerfile:48]: These haven't been reverted yet in charts repo, but they'll stay at 2.0.0 etc...

File: pkg/api/steve/catalog/types/rest.go
Comment (cbron) [inline pkg/api/steve/catalog/types/rest.go:50]: This exists just to adopt the system-charts install right ?

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (thedadams) [inline pkg/controllers/dashboard/hostedkubernetecharts/controller.go:109]: This should a `bool` value. The operators only care that there are additional CAs. They will get them the same way the Rancher chart does.

File: pkg/settings/setting.go
Comment (StrongMonkey) [reply pkg/settings/setting.go:118]: @cbron Yes, min/max can also be used in feature chart, It will fetch the latest chart in the release branch(release-2.6) that is compatible with the min/max setting. The only difference in this model is that for existing setup, we don't upgrade to the latest compatiable model if the min version is satified. So user won't get an upgrade unless they upgrade rancher, which changes min version settings. Fleet uses this model and I was told to apply it to kev2 charts. But if that's not the case I can remove this min version so it always install the latest compatible version which is the same as how it currently works.

File: package/Dockerfile
Comment (StrongMonkey) [reply package/Dockerfile:48]: Oh I see, this can be modified as long as we make a decision on versioning...

File: pkg/api/steve/catalog/types/rest.go
Comment (StrongMonkey) [reply pkg/api/steve/catalog/types/rest.go:50]: Yes, for upgrade case, to adopt resources that were created by helm 2.

**Review by thedadams — CHANGES_REQUESTED** [review]

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (StrongMonkey) [reply pkg/controllers/dashboard/hostedkubernetecharts/controller.go:109]: Oh good catch :+1:

File: pkg/controllers/dashboard/controller.go
Comment (nickgerace) [inline pkg/controllers/dashboard/controller.go:10]: Is this spelling intentional?

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (nickgerace) [inline pkg/controllers/dashboard/hostedkubernetecharts/controller.go:1]: nit: It seems to be intentional. Perhaps, `hostedkubecharts` or `hostedcharts`? I might have missed the conversations leading to this.

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (nickgerace) [inline pkg/controllers/dashboard/hostedkubernetecharts/controller.go:46]: I wonder if we should standardize and/or centralize charts references across our controllers.

- https://github.com/rancher/rancher/blob/a56d10297a277dcc04aa48c6dfd45022dda859e3/pkg/controllers/dashboard/systemcharts/controller.go#L17-L50
- https://github.com/rancher/rancher/blob/a56d10297a277dcc04aa48c6dfd45022dda859e3/pkg/controllers/dashboard/fleetcharts/controller.go#L15-L33

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (nickgerace) [inline pkg/controllers/dashboard/hostedkubernetecharts/controller.go:87]: Might we need to continue if the other chart is not nil?

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (StrongMonkey) [reply pkg/controllers/dashboard/hostedkubernetecharts/controller.go:1]: This was just a random name I gave for the package name, I am up for any good suggestion.

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (StrongMonkey) [reply pkg/controllers/dashboard/hostedkubernetecharts/controller.go:87]: hmm both should be nil given the if condition above...

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (StrongMonkey) [reply pkg/controllers/dashboard/hostedkubernetecharts/controller.go:46]: That's a good one, I am going to standardize it.

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (nickgerace) [reply pkg/controllers/dashboard/hostedkubernetecharts/controller.go:85]: Re-read it. Agreed

File: pkg/controllers/dashboard/hostedkubernetecharts/controller.go
Comment (nickgerace) [reply pkg/controllers/dashboard/hostedkubernetecharts/controller.go:1]: I would personally prefer `hostedkubecharts` or `hostedclustercharts` if we want to keep it similar.

File: pkg/controllers/dashboard/controller.go
Comment (nickgerace) [reply pkg/controllers/dashboard/controller.go:10]: Resolving due to https://github.com/rancher/rancher/pull/33004#discussion_r652315459

**Review by nickgerace — APPROVED** [review]
LGTM, but will likely need reviews from others.

**Comment by rmweir** [comment]
To clarify, I think we want to make sure the operator is as up to date as possible right?

**Comment by thedadams** [comment]
If I have understood the code correctly: if we don't set a min version for the operators in the `Dockerfile`, then the operators would always get updated to the latest version.

**Comment by StrongMonkey** [comment]
@rmweir @thedadams That's correct, if there is no min version defined, the behavior should be alighed with the old system-chart way.

File: pkg/controllers/dashboard/hostedcluster/controller.go
Comment (rmweir) [inline pkg/controllers/dashboard/hostedcluster/controller.go:110]: Is there a reason why ensure needs to be used twice?

Comment (StrongMonkey) [reply pkg/controllers/dashboard/hostedcluster/controller.go:110]: One is for CRD chart and one is for operator chart since these two are separated now.

**Comment by rmweir** [comment]
@StrongMonkey can we remove the min version so they are always updated?

File: package/Dockerfile
Comment (rmweir) [reply package/Dockerfile:48]: @StrongMonkey can we remove the min versions so they are always updated?

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (rmweir) [inline pkg/controllers/management/aks/aks_cluster_handler.go:95]: Nit: Should we handle the scenario where the CRD isn't installed yet more gracefully or wait to proceed until it is installed? Otherwise, errors are likely to occur on L87. We could also return an error saying "waiting for CRD to be created.

Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go:95]: Since the chart operator is based on `v3.cluster`. Once chart is installed it will change v3.cluster which trigger this controller, so it should eventually go through? But yeah I can check if the crd exists and give a more graceful error.

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (rmweir) [inline pkg/controllers/management/aks/aks_cluster_handler.go:257]: I think just return the error is sufficient and setting the message may not be necessary. It's nice but depending on if it is registered or not the expected condition is different here I believe. Also, if we were to update the entire object we probably want to return it from this function and reassign so that the cluster object in the parent function is up to date. This comment applies to all 3 occurrences of this function definition.

**Review by rmweir — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go:257]: Yeah, setting the condition in here so that if crd chart install failed, cluster could have a better way to display the information in UI showing the cluster is waiting on crd. For the cluster object it is passed as a pointer and reassigned in the method, it should get what's updated.

File: pkg/settings/setting.go
Comment (cbron) [reply pkg/settings/setting.go:118]: Seems like you've done that above, do you still need these settings ? 
If its because if you set these you get the "fleet model" there should be comment saying thats how it works, as its somewhat counterintuitive.

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (thedadams) [reply pkg/controllers/management/aks/aks_cluster_handler.go:257]: Reassigning a pointer in a function doesn't bubble up to the caller: https://play.golang.org/p/rdD6A6alGIE

Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go:257]: Oh......

Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go:258]: @rmweir @thedadams This is fixed.

**Review by rmweir — APPROVED** [review]
LGTM after function comment fixes.

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (rmweir) [inline pkg/controllers/management/aks/aks_cluster_handler.go:251]: Can you fix this comment?
```suggestion
// checkCRDReady checks whether necessary CRD, AKSConfig, has been created yet
```

**Review by rmweir — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go]: :+1: fixed

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (thedadams) [inline pkg/controllers/management/aks/aks_cluster_handler.go:252]: This method can be moved to `pkg/controllers/management/clusteroperator/utils.go` and changed to take a string parameter as well: `checkCrdReady(cluster *mgmtv3.Cluster, provider string)`.

Then each handler could pass a string, like: `checkCrdReady(cluster, "aks")`.

We try to keep the common code in `pkg/controllers/management/clusteroperator/utils.go`.

**Review by thedadams — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/aks/aks_cluster_handler.go
Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go:252]: Sounds good, let me change that.

Comment (StrongMonkey) [reply pkg/controllers/management/aks/aks_cluster_handler.go]: :+1: Changed to common function.

**Review by thedadams — DISMISSED** [review]

**Review by rmweir — DISMISSED** [review]
LGTM

File: pkg/settings/setting.go
Comment (cbron) [reply pkg/settings/setting.go:118]: @thedadams @rmweir if we remove these we will have to set min/max on every feature chart version, otherwise every newer release that goes into the release branches will get pulled into old ranchers, which could crash things. Whats your preferred route, updating these all the time or updating min/max in the chart ? 
The chart is better way IMO, but with feature charts its hard to adjust old ones, you have to replace tarball. So we'll either have to go route of: 
a) set a max on every release and continuously update that to bump it
b) not set a max and then add one when there is a new release.
Either ways are error prone. Which then leads me back to using this min version, as it would be obvious if we forgot it.

Comment (cbron) [reply pkg/settings/setting.go:118]: @StrongMonkey I think we should remove these and then we will just put a min on every feature chart version that matches rancher version.

Comment (StrongMonkey) [reply pkg/settings/setting.go]: @cbron Settings are also removed.

**Review by cbron — APPROVED** [review]

**Pull request merged**
